mobile-sdk/src/main/java/com/github/itwin/mobilesdk/jsonvalue/JSONValue.kt

```diff
@@ -0,0 +1,406 @@
+     * [JSONObject.optDouble].
+     */
+    fun optDouble(key: String) = opt(key)?.let { it as? Double ?: (it as? Number)?.toDouble() }
+
+    /**
+     * If the receiver is an array, returns the optional value at the specified index in the array.
+     * If the receiver is not an array, or there is no value at the specified index, returns null.
+     */
+    fun opt(index: Int) = arrayValue?.opt(index)
+}
+
+/**
+ * Returns true if the receiver is a string with a value of "yes", or false otherwise.
+ */
+fun JSONObject.isYes(propertyName: String): Boolean {
+    return optString(propertyName).lowercase() == "yes"
+}
+
+/**
+ * Converts the receiver to a map, converting all [JSONObject] values to maps, all [JSONArray]
+ * values to lists, and all [JSONObject.NULL] values to `null`.
+ */
+fun JSONObject.toMap(): Map<String, Any?> {
+    val result = mutableMapOf<String, Any?>()
+    for (key in keys()) {
+        result[key] = when (val value = this[key]) {
+            is JSONArray    -> value.toList()
+            is JSONObject   -> value.toMap()
+            JSONObject.NULL -> null
+            else            -> value
+        }
+    }
+    return result
+}
+
+/**
+ * Converts the receiver to a list, converting all [JSONObject] values to maps, all [JSONArray]
+ * values to lists, and all [JSONObject.NULL] values to `null`.*
+ */
+fun JSONArray.toList(): List<Any?> {
+    val result = mutableListOf<Any?>()
+    for (i in 0 until length()) {
+        val value = this[i]
+        result.add(when (value) {
+            is JSONArray    -> value.toList()
+            is JSONObject   -> value.toMap()
+            JSONObject.NULL -> null
+            else            -> value
+        })
+    }
+    return result
+}
+
+/**
+ * Returns a new [JSONValue] containing an object value with the specified contents, given as a list
+ * of pairs where the where the first value is the key and the second is the value.
+ *
+ * If multiple pairs have the same key, the resulting object will contain the value from the last
+ * of those pairs.
+ *
+ * __Note__: This is conceptually the same as [mapOf], but creating a [JSONValue] instead of a
+ * [Map].
+ */
+fun jsonOf(vararg pairs: Pair<String, *>): JSONValue {
+    return toJSON(mapOf(*pairs))
+}
+
+/**
+ * Converts a value to a [JSONValue].
+ *
+ * The value must be JSON-compatible. JSON-compatible types are:
+ * * [String]
+ * * [Boolean]
+ * * [Number]
+ * * Array-like type containing JSON-compatible values.
+ * * Object-like type with [String] keys containing JSON-compatible values.
+ *
+ * Array-like types are [List], [Array], and [JSONArray].
+ *
+ * Object-like types are [Map] and [JSONObject].
+ */
+fun toJSON(value: Any?): JSONValue {
+    return when (value) {
+        null         -> JSONValue()
+        is Unit      -> JSONValue(Unit)
+        is Number, is Boolean, is String, is JSONArray, is JSONObject -> JSONValue(value)
```

Comment:
As noted above, this needs to be changed if we add the type-specific constructors.

```suggestion
        is Number     -> JSONValue(value)
        is Boolean    -> JSONValue(value)
        is String     -> JSONValue(value)
        is JSONArray  -> JSONValue(value)
        is JSONObject -> JSONValue(value)
```